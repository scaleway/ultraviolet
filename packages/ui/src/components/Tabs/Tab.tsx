'use client'

import styled from '@emotion/styled'
import type {
  ComponentProps,
  ElementType,
  ForwardedRef,
  KeyboardEventHandler,
  MouseEventHandler,
  ReactNode,
  Ref,
} from 'react'
import { forwardRef, useMemo } from 'react'
import { Badge } from '../Badge'
import { Stack } from '../Stack'
import { Text } from '../Text'
import { Tooltip } from '../Tooltip'
import { useTabsContext } from './TabsContext'

const StyledBadge = styled(Badge)`
  padding: 0 ${({ theme }) => theme.space['1']};
  margin-left: ${({ theme }) => theme.space['1']};
`

const StyledText = styled(Text)``

const StyledTooltip = styled(Tooltip)``

const BadgeContainer = styled.span`
  margin-left: ${({ theme }) => theme.space['1']};
  display: flex;
`

export const StyledTabButton = styled('button')`
  display: flex;
  flex-direction: row;
  padding: ${({ theme }) => `${theme.space['1']} ${theme.space['2']}`};
  cursor: pointer;
  justify-content: center;
  align-items: baseline;
  white-space: nowrap;
  color: ${({ theme }) => theme.colors.neutral.text};
  text-decoration: none;
  user-select: none;
  touch-action: manipulation;
  transition: color 0.2s;
  border: none;
  background: none;
  border-bottom-width: 2px;
  border-bottom-style: solid;
  border-bottom-color: ${({ theme }) => theme.colors.neutral.border};
  outline: none;

  font-size: ${({ theme }) => theme.typography.bodyStrong.fontSize};
  font-family: ${({ theme }) => theme.typography.bodyStrong.fontFamily};
  font-weight: ${({ theme }) => theme.typography.bodyStrong.weight};
  letter-spacing: ${({ theme }) => theme.typography.bodyStrong.letterSpacing};
  line-height: ${({ theme }) => theme.typography.bodyStrong.lineHeight};

  &:hover,
  &:active,
  &:focus {
    text-decoration: none;
    outline: none;
  }

  &:focus-visible {
    outline: auto;
  }

  &[aria-selected='true'] {
    color: ${({ theme }) => theme.colors.primary.text};
    border-bottom-color: ${({ theme }) => theme.colors.primary.border};

    ${StyledText} {
      color: ${({ theme }) => theme.colors.primary.text};
    }
  }

  &[aria-disabled='false']:not(:disabled) {
    &:hover,
    &:focus,
    &:active {
      outline: none;
      color: ${({ theme }) => theme.colors.primary.text};
      border-bottom-color: ${({ theme }) => theme.colors.primary.border};

      &[data-is-selected='false'] {
        ${StyledBadge} {
          background-color: ${({ theme }) => theme.colors.primary.background};
          border-color: ${({ theme }) => theme.colors.primary.background};
          color: ${({ theme }) => theme.colors.primary.text};
        }
        ${StyledText} {
          color: ${({ theme }) => theme.colors.primary.text};
        }
      }
    }
  }

  &[aria-disabled='true'],
  &:disabled {
    cursor: not-allowed;
    filter: grayscale(1) opacity(50%);
  }
`

type TabProps<T extends ElementType = 'button'> = {
  as?: T
  badge?: ReactNode
  children?: ReactNode
  className?: string
  counter?: number | string
  disabled?: boolean
  onClick?: MouseEventHandler<HTMLElement>
  onKeyDown?: KeyboardEventHandler<HTMLElement>
  subtitle?: string
  tooltip?: string
  value?: string | number
} & Omit<
  ComponentProps<T>,
  | 'as'
  | 'badge'
  | 'children'
  | 'className'
  | 'counter'
  | 'disabled'
  | 'value'
  | 'tooltip'
  | 'role'
>

export const Tab = forwardRef(
  <T extends ElementType = 'button'>(
    {
      as,
      badge,
      children,
      className,
      counter,
      disabled = false,
      onClick,
      onKeyDown,
      subtitle,
      tooltip,
      value,
      ...props
    }: TabProps<T>,
    ref: ForwardedRef<HTMLElement>,
  ) => {
    const { selected, onChange } = useTabsContext()
    const computedAs = as ?? 'button'
    const isSelected = useMemo(
      () => value !== undefined && selected === value,
      [value, selected],
    )

    return (
      <StyledTooltip text={tooltip}>
        <StyledTabButton
          role="tab"
          ref={ref as unknown as Ref<HTMLButtonElement>}
          className={className}
          as={computedAs}
          aria-label={value ? `${value}` : undefined}
          aria-selected={isSelected}
          aria-disabled={disabled}
          disabled={computedAs === 'button' ? disabled : undefined}
          type={computedAs === 'button' ? 'button' : undefined}
          onClick={event => {
            if (value !== undefined) {
              onChange(value)
            }
            onClick?.(event)
          }}
          onKeyDown={event => {
            onKeyDown?.(event)
            if (!event.defaultPrevented && !disabled && value) onChange(value)
          }}
          data-is-selected={isSelected}
          {...props}
        >
          <Stack direction="column" gap={0.5}>
            <Stack direction="row" alignItems="center">
              {children}
              {typeof counter === 'number' || typeof counter === 'string' ? (
                <StyledBadge
                  sentiment={isSelected ? 'primary' : 'neutral'}
                  prominence={isSelected ? 'strong' : 'default'}
                  size="medium"
                >
                  {counter}
                </StyledBadge>
              ) : null}
              {badge ? <BadgeContainer>{badge}</BadgeContainer> : null}
            </Stack>
            {subtitle ? (
              <Stack direction="row">
                <StyledText
                  as="span"
                  variant="bodySmall"
                  sentiment="neutral"
                  prominence="weak"
                >
                  {subtitle}
                </StyledText>
              </Stack>
            ) : null}
          </Stack>
        </StyledTabButton>
      </StyledTooltip>
    )
  },
)
