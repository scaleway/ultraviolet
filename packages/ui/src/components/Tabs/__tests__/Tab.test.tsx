import { shouldMatchEmotionSnapshot } from '@utils/test'
import { describe, test } from 'vitest'
import { Tab } from '../Tab'

describe('Tab', () => {
  test('renders correctly', () => shouldMatchEmotionSnapshot(<Tab />))
  test('renders correctly with subtitle', () =>
    shouldMatchEmotionSnapshot(<Tab subtitle="test" />))

  test('renders correctly with counter, badge and subtitle', () =>
    shouldMatchEmotionSnapshot(
      <Tab badge="badge" counter={1} subtitle="subtitle" />,
    ))
})
