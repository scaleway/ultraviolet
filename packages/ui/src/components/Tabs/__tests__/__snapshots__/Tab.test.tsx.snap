// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tab > renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding: 0.5rem 1rem;
  cursor: pointer;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  white-space: nowrap;
  color: #3f4250;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  -webkit-transition: color 0.2s;
  transition: color 0.2s;
  border: none;
  background: none;
  border-bottom-width: 2px;
  border-bottom-style: solid;
  border-bottom-color: #d9dadd;
  outline: none;
  font-size: 1rem;
  font-family: Inter,sans-serif;
  font-weight: 500;
  letter-spacing: 0;
  line-height: 1.5rem;
}

.emotion-0:hover,
.emotion-0:active,
.emotion-0:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: none;
}

.emotion-0:focus-visible {
  outline: auto;
}

.emotion-0[aria-selected='true'] {
  color: #641cb3;
  border-bottom-color: #8c40ef;
}

.emotion-0[aria-selected='true'] .e1hzf7cr3 {
  color: #641cb3;
}

.emotion-0[aria-disabled='false']:not(:disabled):hover,
.emotion-0[aria-disabled='false']:not(:disabled):focus,
.emotion-0[aria-disabled='false']:not(:disabled):active {
  outline: none;
  color: #641cb3;
  border-bottom-color: #8c40ef;
}

.emotion-0[aria-disabled='false']:not(:disabled):hover[data-is-selected='false'] .e1hzf7cr4,
.emotion-0[aria-disabled='false']:not(:disabled):focus[data-is-selected='false'] .e1hzf7cr4,
.emotion-0[aria-disabled='false']:not(:disabled):active[data-is-selected='false'] .e1hzf7cr4 {
  background-color: #f1eefc;
  border-color: #f1eefc;
  color: #641cb3;
}

.emotion-0[aria-disabled='false']:not(:disabled):hover[data-is-selected='false'] .e1hzf7cr3,
.emotion-0[aria-disabled='false']:not(:disabled):focus[data-is-selected='false'] .e1hzf7cr3,
.emotion-0[aria-disabled='false']:not(:disabled):active[data-is-selected='false'] .e1hzf7cr3 {
  color: #641cb3;
}

.emotion-0[aria-disabled='true'],
.emotion-0:disabled {
  cursor: not-allowed;
  -webkit-filter: grayscale(1) opacity(50%);
  filter: grayscale(1) opacity(50%);
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.25rem;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

<div
    data-testid="testing"
  >
    <button
      aria-disabled="false"
      aria-selected="false"
      class="emotion-0 emotion-1"
      data-is-selected="false"
      role="tab"
      type="button"
    >
      <div
        class="emotion-2 emotion-3"
      >
        <div
          class="emotion-4 emotion-3"
        />
      </div>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Tab > renders correctly with counter, badge and subtitle 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding: 0.5rem 1rem;
  cursor: pointer;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  white-space: nowrap;
  color: #3f4250;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  -webkit-transition: color 0.2s;
  transition: color 0.2s;
  border: none;
  background: none;
  border-bottom-width: 2px;
  border-bottom-style: solid;
  border-bottom-color: #d9dadd;
  outline: none;
  font-size: 1rem;
  font-family: Inter,sans-serif;
  font-weight: 500;
  letter-spacing: 0;
  line-height: 1.5rem;
}

.emotion-0:hover,
.emotion-0:active,
.emotion-0:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: none;
}

.emotion-0:focus-visible {
  outline: auto;
}

.emotion-0[aria-selected='true'] {
  color: #641cb3;
  border-bottom-color: #8c40ef;
}

.emotion-0[aria-selected='true'] .emotion-13 {
  color: #641cb3;
}

.emotion-0[aria-disabled='false']:not(:disabled):hover,
.emotion-0[aria-disabled='false']:not(:disabled):focus,
.emotion-0[aria-disabled='false']:not(:disabled):active {
  outline: none;
  color: #641cb3;
  border-bottom-color: #8c40ef;
}

.emotion-0[aria-disabled='false']:not(:disabled):hover[data-is-selected='false'] .emotion-7,
.emotion-0[aria-disabled='false']:not(:disabled):focus[data-is-selected='false'] .emotion-7,
.emotion-0[aria-disabled='false']:not(:disabled):active[data-is-selected='false'] .emotion-7 {
  background-color: #f1eefc;
  border-color: #f1eefc;
  color: #641cb3;
}

.emotion-0[aria-disabled='false']:not(:disabled):hover[data-is-selected='false'] .emotion-13,
.emotion-0[aria-disabled='false']:not(:disabled):focus[data-is-selected='false'] .emotion-13,
.emotion-0[aria-disabled='false']:not(:disabled):active[data-is-selected='false'] .emotion-13 {
  color: #641cb3;
}

.emotion-0[aria-disabled='true'],
.emotion-0:disabled {
  cursor: not-allowed;
  -webkit-filter: grayscale(1) opacity(50%);
  filter: grayscale(1) opacity(50%);
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.25rem;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-7 {
  font-size: 0.75rem;
  font-family: Inter,sans-serif;
  font-weight: 400;
  letter-spacing: 0;
  line-height: 1rem;
  text-transform: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  padding: 0 0.5rem;
  margin-left: 0.5rem;
}

.emotion-9 {
  margin-left: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-14 {
  color: #727683;
  font-size: 0.875rem;
  font-family: Inter,sans-serif;
  font-weight: 400;
  letter-spacing: 0;
  line-height: 1.25rem;
  text-transform: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
    data-testid="testing"
  >
    <button
      aria-disabled="false"
      aria-selected="false"
      class="emotion-0 emotion-1"
      data-is-selected="false"
      role="tab"
      type="button"
    >
      <div
        class="emotion-2 emotion-3"
      >
        <div
          class="emotion-4 emotion-3"
        >
          <span
            class="emotion-6 styles__1yf71jy0 styles_size_medium__1yf71jy2 styles_prominence_default__1yf71jyd styles_sentiment_neutral__1yf71jya styles_undefined_compound_6__1yf71jym emotion-7 emotion-8"
          >
            1
          </span>
          <span
            class="emotion-9 emotion-10"
          >
            badge
          </span>
        </div>
        <div
          class="emotion-11 emotion-3"
        >
          <span
            class="emotion-13 emotion-14 emotion-8"
          >
            subtitle
          </span>
        </div>
      </div>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Tab > renders correctly with subtitle 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding: 0.5rem 1rem;
  cursor: pointer;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  white-space: nowrap;
  color: #3f4250;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  -webkit-transition: color 0.2s;
  transition: color 0.2s;
  border: none;
  background: none;
  border-bottom-width: 2px;
  border-bottom-style: solid;
  border-bottom-color: #d9dadd;
  outline: none;
  font-size: 1rem;
  font-family: Inter,sans-serif;
  font-weight: 500;
  letter-spacing: 0;
  line-height: 1.5rem;
}

.emotion-0:hover,
.emotion-0:active,
.emotion-0:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: none;
}

.emotion-0:focus-visible {
  outline: auto;
}

.emotion-0[aria-selected='true'] {
  color: #641cb3;
  border-bottom-color: #8c40ef;
}

.emotion-0[aria-selected='true'] .emotion-9 {
  color: #641cb3;
}

.emotion-0[aria-disabled='false']:not(:disabled):hover,
.emotion-0[aria-disabled='false']:not(:disabled):focus,
.emotion-0[aria-disabled='false']:not(:disabled):active {
  outline: none;
  color: #641cb3;
  border-bottom-color: #8c40ef;
}

.emotion-0[aria-disabled='false']:not(:disabled):hover[data-is-selected='false'] .e1hzf7cr4,
.emotion-0[aria-disabled='false']:not(:disabled):focus[data-is-selected='false'] .e1hzf7cr4,
.emotion-0[aria-disabled='false']:not(:disabled):active[data-is-selected='false'] .e1hzf7cr4 {
  background-color: #f1eefc;
  border-color: #f1eefc;
  color: #641cb3;
}

.emotion-0[aria-disabled='false']:not(:disabled):hover[data-is-selected='false'] .emotion-9,
.emotion-0[aria-disabled='false']:not(:disabled):focus[data-is-selected='false'] .emotion-9,
.emotion-0[aria-disabled='false']:not(:disabled):active[data-is-selected='false'] .emotion-9 {
  color: #641cb3;
}

.emotion-0[aria-disabled='true'],
.emotion-0:disabled {
  cursor: not-allowed;
  -webkit-filter: grayscale(1) opacity(50%);
  filter: grayscale(1) opacity(50%);
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.25rem;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

<div
    data-testid="testing"
  >
    <button
      aria-disabled="false"
      aria-selected="false"
      class="emotion-0 emotion-1"
      data-is-selected="false"
      role="tab"
      type="button"
    >
      <div
        class="emotion-2 emotion-3"
      >
        <div
          class="emotion-4 emotion-3"
        />
        <div
          class="emotion-6 emotion-3"
        >
          <span
            class="emotion-8 emotion-9 style__m4c9ow0 style_strikeThrough_false__m4c9ow2 style_italic_false__m4c9ow4 style_underline_false__m4c9ow6 style_oneLine_false__m4c9ow8 style_sentiment_neutral__m4c9owb style_prominence_weak__m4c9owl style_variant_bodySmall__m4c9own style_disabled_false__m4c9ow1b style_undefined_compound_27__m4c9ow23 style_undefined_compound_75__m4c9ow3f"
          >
            test
          </span>
        </div>
      </div>
    </button>
  </div>
</DocumentFragment>
`;
